### ISSUE TITLE
get_global in data segment

### ISSUE THREAD

**Issue by pannous** [WebAssembly/binaryen#1477]
According to the specification, a data segment wants a constant-expression, which `(get_global $hello)` is:
https://webassembly.github.io/spec/core/valid/instructions.html#valid-constant

So why does the second variant not work here: ?
```
 (global $hello i32 (i32.const 1))
 (data (i32.const 1) "hello wasm!\00") ;; works, but no identifier here :(
 (data (get_global $hello) "Hello global!\00") ;; is this supposed to work?
```

**Comment by kripken** [comment]
Yeah, I think you're right, it looks valid according to the spec. Binaryen appears to allow it, but wabt errors with
```
a.wast:6:9: error: initializer expression can only reference an imported global
```
@binji, thoughts?

**Comment by binji** [comment]
The spec interpreter has the same error:
```
$ cat test.wat
(global $hello i32 (i32.const 1))
(memory 1)
(data (get_global $hello) "hi")

$ ./wasm test.wat
test.wat:3.19-3.25: invalid module: unknown global 0
```

But if you import the global it works:
```
$ cat test2.wat 
(global $hello (import "spectest" "global") i32)
(memory 1)
(data (get_global $hello) "hi")

$ ./wasm test2.wat
```

**Comment by binji** [comment]
There's even [a test](https://github.com/WebAssembly/spec/blob/7b7911eba79c06fe0511f4af65771c8e4325ebf4/test/core/data.wast#L72-L74) for this case.

Hm, though now that you mention it, that rule makes sense for preventing recursive global definitions (and is defined in the spec [here](https://webassembly.github.io/spec/core/exec/modules.html#instantiation)), but that uses a different frame `F_im` for evaluation (see step 5.a and 5.b).

But the offset evaluation should be using `F` instead (pushed in step 8).


@rossberg, am I missing something, or is this a bug in the spec?

**Comment by tlively** [comment]
MVP Wasm only allowed constant imported globals to be used in constant expressions. The GC proposal extends this to include previously defined constant globals as well. Both are implemented in Binaryen.
